File: chapter-03/03-late-data/02-sequential-late-data-integrator-version-delta-lake/backfilling-configurator/src/main/scala/com/waitingforcode/SparkSessionFactory.scala
package com.waitingforcode

import io.delta.sql.DeltaSparkSessionExtension
import org.apache.spark.sql.SparkSession

class SparkSessionFactory(outputLocation: String) {


  val derbyLocation = s"${outputLocation}/derby/"

  val sparkSession: SparkSession = {
    SparkSession.builder()
      .config("spark.sql.session.timeZone", "UTC")
      .withExtensions(new DeltaSparkSessionExtension())
      .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog")
      .config("spark.sql.warehouse.dir", s"${outputLocation}/spark/warehouse")
      .enableHiveSupport()
      .getOrCreate()
  }

}

object SparkSessionFactoryInitializer {

  def init(outputLocation: String): SparkSessionFactory = {
    val sparkSessionFactory = new SparkSessionFactory(outputLocation)
    System.setProperty("derby.system.home", sparkSessionFactory.derbyLocation)
    sparkSessionFactory
  }

}

File: chapter-10/03-data-lineage/02-fine-grained-tracker-apache-spark-openlineage-marquez/spark_session_factory.py
from delta import configure_spark_with_delta_pip
from pyspark.sql import SparkSession


def create_spark_session_with_delta_lake_and_open_lineage(app_name: str) -> SparkSession:
    return configure_spark_with_delta_pip(SparkSession.builder.master("local[*]")
                                          .appName(app_name)
                                          .config("spark.sql.catalogImplementation", "hive")
                                          .config("spark.sql.extensions",
                                                  "io.delta.sql.DeltaSparkSessionExtension")
                                          .config("spark.sql.catalog.spark_catalog",
                                                  "org.apache.spark.sql.delta.catalog.DeltaCatalog")
                                          .config("spark.extraListeners",
                                                  "io.openlineage.spark.agent.OpenLineageSparkListener")
                                          .config("spark.openlineage.transport.type", "http")
                                          .config("spark.openlineage.transport.url", "http://localhost:5000")
                                          .config("spark.openlineage.namespace", "visits"),
                                          extra_packages=['io.openlineage:openlineage-spark_2.12:1.21.1']
                                          ).getOrCreate()
